File: problems/2109. Adding Spaces to a String/python3/main.py
from typing import List


class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        newstr = [""] * (len(s) + len(spaces))
        new_i = 0
        s_i = 0

        for i in range(len(s)):
            # print(f"new_i-{new_i}, i-{i} | {newstr}")
            while len(spaces) > s_i and new_i - s_i == spaces[s_i]:
                newstr[new_i] = " "
                s_i += 1
                new_i += 1

            newstr[new_i] = s[i]
            new_i += 1

        # print("".join(newstr))
        return "".join(newstr)


if __name__ == "__main__":

    assert (
        Solution().addSpaces("LeetcodeHelpsMeLearn", [8, 13, 15])
        == "Leetcode Helps Me Learn"
    )

File: problems/2109. Adding Spaces to a String/cpp/main.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    string addSpaces(string s, vector<int>& spaces) {
		string out;
		// out.reserve(s.size() + spaces.size());

		int  idx = 0;

		for (int i = 0; i < s.size(); i++) {
			if (idx < spaces.size() && i == spaces[idx]) {
				out += " ";
				idx++;
			}
			out += s[i];
		}

		return out;
    }
};


int main() {
	auto s = Solution();
	vector<int> spaces;

	spaces = {8, 13, 15};
	assert(s.addSpaces("LeetcodeHelpsMeLearn", spaces) ==  "Leetcode Helps Me Learn");

	spaces = {1,5,7,9};
	assert(s.addSpaces("icodeinpython", spaces) == "i code in py thon");

	spaces = {0,1,2,3,4,5,6};
	assert(s.addSpaces("spacing", spaces) == " s p a c i n g");
}
